src/plugins/action_increment_stage/src/incrementStage.ts

```diff
@@ -0,0 +1,93 @@
+import { ActionPluginOutput } from '../../types'
+
+type IParameters = {
+  applicationId: number
+}
+
+type Stage = {
+  id: number
+  number: number
+  title: string
+}
+
+module.exports['incrementStage'] = async function (
+  parameters: IParameters,
+  DBConnect: any
+): Promise<ActionPluginOutput> {
+  const { applicationId } = parameters
+  const returnObject: ActionPluginOutput = { status: null, error_log: '' }
+  console.log(`Incrementing the Stage for Application ${applicationId}...`)
+
+  try {
+    const templateId: number = await DBConnect.getTemplateId('application', applicationId)
```

Comment:
I think the logic here can be simplified a bit, following this:

`getCurrentStatusHistory` (add number of application_stage_history to query)

if status is not undefined, create completed (should have everything for that from previous query), and set next status = draft, otherwise next status = status.

`getNextStageID` new SQL
```
select min(stageID) from template_stage
 where (number >  {number from `getCurrentStatusHistory` or 0} and template_id = {templateId} } 
```
if stageID undefined, then its last stage, return (can add some msg to returnObject)

add stage
add status

I personally think early returns of returnObject (on failure etc..) is much easier to follow

Reply:
Thanks for this suggestion. I've refactored it quite a bit and it's a lot clearer. The new VIEW which includes all stage_status data combined (including IDs, which the old one lacked) makes it simpler as well.

Will push shortly :)